### PR TITLE
feat: Import all SMT files into cosmos-sdk without any build errors

### PR DESCRIPTION
**Pull request by Manav-Aggarwal**
## Description

Closes: https://github.com/celestiaorg/cosmos-sdk/issues/231

For Cevmos' Block Fraud Proofs we need to have a [SMT](https://github.com/celestiaorg/smt) based store. We want to try and backport the [one from v0.46.x](https://github.com/vulcanize/cosmos-sdk/releases/tag/v0.46.0-smt-0.0.5-alpha). 

Follow up would be in #https://github.com/celestiaorg/cosmos-sdk/issues/111

Please use https://github.com/celestiaorg/cosmos-sdk/compare/manav/backport_smt-add_files...vulcanize:cosmos-sdk:release-v0.46.0-smt-0.0.5-alpha for comparing this to the vulcanize implementation if useful for review

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by liamsi — COMMENTED** [review]
It looks like the bulk changes here are from https://github.com/vulcanize/cosmos-sdk/releases/tag/v0.46.0-smt-0.0.5-alpha?

Can we instead cherry-pick changes and then resolve merge conflicts? Even if github does not give any real guarantees about the commits being the same it will help reviewing the PR a bit IMO.

**Comment by jbowen93** [comment]
> It looks like the bulk changes here are from https://github.com/vulcanize/cosmos-sdk/releases/tag/v0.46.0-smt-0.0.5-alpha?
> 
> Can we instead cherry-pick changes and then resolve merge conflicts? Even if github does not give any real guarantees about the commits being the same it will help reviewing the PR a bit IMO.

Would you prefer cherry picking the commits from Vulkanize and then layering commits on top of that to resolve changes needed for v0.46 -> v0.45?

**Comment by Manav-Aggarwal** [comment]
Looks like there's a race condition in a test sometimes. Will move on to #111 and build on top of this for now

**Comment by Manav-Aggarwal** [comment]
For review: https://github.com/celestiaorg/cosmos-sdk/compare/manav/backport_smt-add_files...vulcanize:cosmos-sdk:release-v0.46.0-smt-0.0.5-alpha

**Comment by Manav-Aggarwal** [comment]
> It looks like the bulk changes here are from https://github.com/vulcanize/cosmos-sdk/releases/tag/v0.46.0-smt-0.0.5-alpha?
> 
> Can we instead cherry-pick changes and then resolve merge conflicts? Even if github does not give any real guarantees about the commits being the same it will help reviewing the PR a bit IMO.

Please correct me if I am wrong but wouldn't that mean looking at https://github.com/celestiaorg/cosmos-sdk/compare/manav/backport_smt...vulcanize:cosmos-sdk:release-v0.46.0-smt-0.0.5-alpha and cherry-picking commits? It seems like there are 1k+ commits there which might not be worth the effort

**Review by adlerjohn — COMMENTED** [review]
Can you rebase `manav/backport_smt` against the latest default branch? As-is, this PR is useless since it's based on a stale Cosmos SDK release.

File: .github/workflows/labeler.yml [hunk 1]
```diff
@@ -5,6 +5,9 @@ on:
 jobs:
   labeler:
     runs-on: ubuntu-latest
+    permissions:
+      issues: write
+      pull-requests: write
```
Comment (adlerjohn) [inline .github/workflows/labeler.yml:10]: Why was this changed?

Comment (Manav-Aggarwal) [reply .github/workflows/labeler.yml]: was changed as an attempt to get the `Pull Request Labeler` test to pass but it seems like an issue on the repo level. Reverted for now

File: store/cachemulti/store.go [hunk 2]
```diff
@@ -23,8 +24,9 @@ type Store struct {
 	stores map[types.StoreKey]types.CacheWrap
 	keys   map[string]types.StoreKey
 
-	traceWriter  io.Writer
-	traceContext types.TraceContext
+	traceWriter       io.Writer
+	traceContext      types.TraceContext
+	traceContextMutex sync.RWMutex
```
Comment (evan-forbes) [inline store/cachemulti/store.go:29]: should the mutex be a pointer?

**Comment by Manav-Aggarwal** [comment]
> Can you rebase `manav/backport_smt` against the latest default branch? As-is, this PR is useless since it's based on a stale Cosmos SDK release.

Rebased to 0.45.5 which is what ethermint uses

File: store/cachemulti/store.go (hunk 2 above)
Comment (Manav-Aggarwal) [reply store/cachemulti/store.go:29]: Just rebased on top of 0.45.5 instead, and that seems to have fixed the issue. Thanks Evan! Appreciate your help :)

**Comment by liamsi** [comment]
What is the status of this PR?

**Comment by Manav-Aggarwal** [comment]
> What is the status of this PR? Ethermint now uses 0.46

good from my side, waiting on review

**Comment by liamsi** [comment]
I don't really follow. I thought this backports code from 0.46 to an older version of the SDK? Why would that still be necessary when you'd be now using 0.46 directly? Are you sure this is good to go as is @Manav-Aggarwal?

**Comment by liamsi** [comment]
Also, related to this:
> Please correct me if I am wrong but wouldn't that mean looking at https://github.com/celestiaorg/cosmos-sdk/compare/manav/backport_smt...vulcanize:cosmos-sdk:release-v0.46.0-smt-0.0.5-alpha and cherry-picking commits? It seems like there are 1k+ commits there which might not be worth the effort

You should compare the diff of their work from where they forked off and not against your branch in our fork. If you look at 
https://github.com/cosmos/cosmos-sdk/compare/main...vulcanize:cosmos-sdk:release-v0.46.0-smt-0.0.5-alpha you can see  they are only 86 commits, and mostly related to the SMT stuff the vlucanize team has been working on only.

**Comment by Manav-Aggarwal** [comment]
> I don't really follow. I thought this backports code from 0.46 to an older version of the SDK? Why would that still be necessary when you'd be now using 0.46 directly? Are you sure this is good to go as is @Manav-Aggarwal?

Since cosmos-sdk v0.46 supports tendermint v0.34 now, this PR is not needed. We'll be closing this PR and prioritizing https://github.com/celestiaorg/cosmos-sdk/issues/245 directly in cosmos-sdk 0.46
